neo/Consensus/ConsensusContext.cs

```diff
@@ -167,13 +180,37 @@ private ConsensusPayload MakeSignedPayload(ConsensusMessage message)
                 PrevHash = PrevHash,
                 BlockIndex = BlockIndex,
                 ValidatorIndex = (ushort)MyIndex,
-                Timestamp = Timestamp,
-                Data = message.ToArray()
+                Timestamp = (overrideTimestamp != 0) ? overrideTimestamp : Timestamp,
```

Comment:
Why do we have to override the `Timestamp`?

Reply:
Because for `ChangeView` I set it equal to the time the `ChangeView` message is created. I have a comment about why for protecting against replay.

Reply:
https://github.com/neo-project/neo/blob/5b28ca4aba244814381ed63ab9b79f79ed44379d/neo/Consensus/ConsensusContext.cs#L134-L136

Reply:
Moreover, NEO's p2p messages themselves can prevent replay attacks because a node does not receive messages with the same hash.

https://github.com/neo-project/neo/blob/7cf286941f7f5541e2ab4bf182583452eec8fc5a/neo/Network/P2P/TaskManager.cs#L59

Reply:
It happens in normal operation too. This improves the performance in normal operation.

Reply:
As the network gets bigger this happens even more in normal operation.

Reply:
Something needs to change the hash whether it is blocked at a higher level or here. If the hash doesn't change for different unique request for the same view then there isn't  way to block it.

Reply:
In normal operation getdata does not broadcast it's response to all nodes.

Reply:
I put in some logging to see that this was getting hit, and it was happening even in a small situation of 4 consensus nodes and a couple other nodes relaying messages.

Reply:
In fact, each node will only request once for the same view number.

Reply:
> https://github.com/neo-project/neo/blob/7cf286941f7f5541e2ab4bf182583452eec8fc5a/neo/Network/P2P/TaskManager.cs#L59
This does not apply to these consensus messages:

Reply:
> In fact, each node will only request once for the same view number.

Yes but it will receive that message many times. The code in TaskManager does not apply to these messages.

Reply:
If you override the `Timestamp`, you will still receive that message many times.

Reply:
It's not as big of a problem for other messages because if they receive the same one twice they don't respond, but for `ChangeView` messages it is a problem now without this, because it now respondes with `RecoveryMessage` and would have needlessly poor performance with a large network without this mechanism.

Reply:
> If you override the `Timestamp`, you will still receive that message many times.

Yes but you have a different hash so you don't have to respond many times!

Reply:
Because when it responds it broadcasts to all nodes. !

Reply:
This code will see it has the same hash and only respond once.

Reply:
I do not quite understand. If a node can only request the same view number once, then whether or not to override the `Timestamp`, you can use hash to prevent replay.

Reply:
No you can't because if you use the hash to prevent replay; and the node restarts and sends the request again; in that case you do want to respond, but it will have the same timestamp, and thus not respond.

Reply:
Node requests change view 1, then 2, then 3.
Restarts:
requests change view 1, 2, 3 a second time; would have the same hash and receivers wouldn't respond again.

Reply:
This solution allows actual new requests for the same view number to get a response again, while allowing blocking duplicate replays.

Reply:
Perhaps it should block it at a higher level if it has the same hash, but that makes it even more important that it have unique hashes for separate requests.

Reply:
There could be an additional timestamp in all the messages for the time the message was sent if you don't like overriding the use of this timestamp.

Reply:
Okay, I understand. Then I think there are two ways to solve this problem:

1. As you can, add an additional timestamp to the `ChangeView` message.
2. `RecoveryMessage` is not requested through the `ChangeView` message, but by adding a `RecoveryRequest` message.

Reply:
In addition, it seems that there is no need to put `Timestamp` in `ConsensusPayload`. I don't remember why I designed this in the first place, but maybe moving `Timestamp` to `PrepareRequest` is a better way.

Reply:
Hmm, that sounds like a good idea. Let me pursue that.

Reply:
I think I'm going to be able to simplify other things in doing this also.

Reply:
The `PrepareRequest` can be directly composed into the `RecoveryMessage` now.

Reply:
This has been addressed, the `Timestamp` has been moved out of the `ConsensusPayload`.